src/integrationTest/java/com/hivemq/client/restrictions/Mqtt5ToMqtt3MessageForwardingIT.java

```diff
@@ -0,0 +1,221 @@
+                                            .builder()
+                                            .topic("test_1")
+                                            .payload(payload)
+                                            .qos(MqttQos.EXACTLY_ONCE)
+                                            .build()
+                            )
+                    )
+                    .timeout(5, TimeUnit.SECONDS)
+                    .subscribe();
+
+            TimeUnit.SECONDS.sleep(1);
+        }
+
+        assertTrue(countDownLatch.await(5 + 1, TimeUnit.SECONDS));
+    }
+
+    @Test
+    void mqtt5ToMqtt3_messageForwarding_async() throws InterruptedException, ExecutionException, TimeoutException {
+
+        final Mqtt3AsyncClient mqtt3Client = MqttClient
+                .builder()
+                .useMqttVersion3()
+                .serverPort(hivemq.getMqttPort())
+                .addConnectedListener(__ -> System.out.println("MQTTv3 client connected."))
+                .addDisconnectedListener(context -> System.out.println("MQTTv3 client disconnected. (" + context.getCause().getMessage() + ")"))
+                .buildAsync();
+
+        final Mqtt5AsyncClient mqtt5Client = MqttClient
+                .builder()
+                .useMqttVersion5()
+                .serverPort(hivemq.getMqttPort())
+                .addConnectedListener(__ -> System.out.println("MQTTv5 client connected."))
+                .addDisconnectedListener(context -> System.out.println("MQTTv5 client disconnected. (" + context.getCause().getMessage() + ")"))
+                .buildAsync();
+
+        final Mqtt3ConnAck mqtt3ConnAck = mqtt3Client.connect().get(5, TimeUnit.SECONDS);
+        assertEquals(Mqtt3ConnAckReturnCode.SUCCESS, mqtt3ConnAck.getReturnCode());
+
+        final Mqtt5ConnAck mqtt5ConnAck = mqtt5Client.connect().get(5, TimeUnit.SECONDS);
+        assertEquals(Mqtt5ConnAckReasonCode.SUCCESS, mqtt5ConnAck.getReasonCode());
+
+        final int MESSAGE_COUNT = 10;
+        final CountDownLatch countDownLatch = new CountDownLatch(MESSAGE_COUNT);
+
+        final Mqtt5SubAck mqtt5SubAck = mqtt5Client
+                .subscribeWith()
+                .topicFilter("test_1")
+                .qos(MqttQos.EXACTLY_ONCE)
+                .callback(message -> {
+                    final byte[] payload = message.getPayloadAsBytes();
+                    final int i = payload[0];
+                    System.out.println("MQTTv5 client received message #" + i + " from topic \"test_1\".");
+                    System.out.println("MQTTv3 client sending message #" + i + " to topic \"test_2\"...");
+                    try {
+                        mqtt3Client
+                                .publishWith()
+                                .topic("test_2")
+                                .payload(payload)
+                                .qos(MqttQos.EXACTLY_ONCE)
+                                .send()
+                                .get(5, TimeUnit.SECONDS);
```

Comment:
This code blocks the thread that calls the callback. Blocking inside the callback is not allowed.

Reply:
The acknowledgement can actually not be delivered when it wants to use the same thread that you are currently blocking. This creates a sort of deadlock here.

Reply:
As you use a QoS > 0, you should use manual acknowledgement and acknowledge the message received by the MQTT 5 client when successfully delivered with the MQTT 3 client. Blocking in the callback does not give you guaranteed message delivery (and is not allowed here anyhow).

Reply:
Thanks for looking into this, @SgtSilvio.

Is there documentation for this restriction? I don't get why the acknowledgment would not be able to be delivered in this case. I mean, it blocks because it is waiting on the acknowledgement, doesn't it? The fact that calling the blocking client inside the callback works makes it extra confusing to understand what you are saying:

```java
// Passing test:
@Test
void mqtt5ToMqtt3_messageForwarding_blocking_in_async() throws InterruptedException, ExecutionException, TimeoutException {

    final Mqtt3BlockingClient mqtt3Client = MqttClient
            .builder()
            .useMqttVersion3()
            .serverPort(hivemq.getMqttPort())
            .addConnectedListener(__ -> System.out.println("MQTTv3 client connected."))
            .addDisconnectedListener(context -> System.out.println("MQTTv3 client disconnected. (" + context.getCause().getMessage() + ")"))
            .buildBlocking();

    final Mqtt5AsyncClient mqtt5Client = MqttClient
            .builder()
            .useMqttVersion5()
            .serverPort(hivemq.getMqttPort())
            .addConnectedListener(__ -> System.out.println("MQTTv5 client connected."))
            .addDisconnectedListener(context -> System.out.println("MQTTv5 client disconnected. (" + context.getCause().getMessage() + ")"))
            .buildAsync();

    final Mqtt3ConnAck mqtt3ConnAck = mqtt3Client.connect();
    assertEquals(Mqtt3ConnAckReturnCode.SUCCESS, mqtt3ConnAck.getReturnCode());

    final Mqtt5ConnAck mqtt5ConnAck = mqtt5Client.connect().get(5, TimeUnit.SECONDS);
    assertEquals(Mqtt5ConnAckReasonCode.SUCCESS, mqtt5ConnAck.getReasonCode());

    final int MESSAGE_COUNT = 10;
    final CountDownLatch countDownLatch = new CountDownLatch(MESSAGE_COUNT);

    final Mqtt5SubAck mqtt5SubAck = mqtt5Client
            .subscribeWith()
            .topicFilter("test_1")
            .qos(MqttQos.EXACTLY_ONCE)
            .callback(message -> {
                final byte[] payload = message.getPayloadAsBytes();
                final int i = payload[0];
                System.out.println("MQTTv5 client received message #" + i + " from topic \"test_1\".");
                System.out.println("MQTTv3 client sending message #" + i + " to topic \"test_2\"...");
                try {
                    mqtt3Client
                            .publishWith()
                            .topic("test_2")
                            .payload(payload)
                            .qos(MqttQos.EXACTLY_ONCE)
                            .send();

                    countDownLatch.countDown();
                } catch (final Exception ex) {
                    System.out.println("MQTTv3 client failed to send message #" + i + ".");
                    ex.printStackTrace();
                }
            })
            .send()
            .get(5, TimeUnit.SECONDS);

    assertTrue(mqtt5SubAck.getReasonCodes().contains(Mqtt5SubAckReasonCode.GRANTED_QOS_2));

    for (int i = 0; i < MESSAGE_COUNT; i++) {
        final byte[] payload = new byte[42];
        payload[0] = (byte) i;

        System.out.println("MQTTv5 client sending message #" + i + " to topic \"test_1\"...");
        mqtt5Client
                .publishWith()
                .topic("test_1")
                .payload(payload)
                .qos(MqttQos.EXACTLY_ONCE)
                .send()
                .get(5, TimeUnit.SECONDS);

        TimeUnit.SECONDS.sleep(1);
    }

    assertTrue(countDownLatch.await(5 + 1, TimeUnit.SECONDS));
}
```

To be honest, I have always assumed that the callback consumers were executed in separate threads from a thread pool. You saying this now at least helps me understand why throwing an exception in a callback messes up with all subsequent calls (https://github.com/hivemq/hivemq-mqtt-client/issues/575) – there really is only one thread doing the work underneath, isn't there?

Reply:
> there really is only one thread doing the work underneath
no, there are multiple threads, otherwise it would happen right on the first message.

Reply:
I will explain it in more detail, but today is a full day, so will take some time